mindmeld/system_entity_recognizer.py

```diff
@@ -41,61 +50,194 @@ class SystemEntityError(Exception):
+        In general we want to find the system entity recognizer from the application config.
+        If the application configuration is empty, we do not use Duckling.
+        Otherwise, we return the Duckling recognizer with the URL defined in the application's
+          config, default to the DEFAULT_DUCKLING_URL.
+
+        Args:
+            app_path: The application path
+
+        Returns:
+            (SystemEntityRecognizer): A SystemEntityRecognizer instance
+        """
+        if not SystemEntityRecognizer._instance:
+            if is_duckling_configured(app_path):
+                url = get_system_entity_url_config(app_path=app_path)
+                SystemEntityRecognizer._instance = DucklingRecognizer.get_instance(url)
+            else:
+                SystemEntityRecognizer._instance = (
+                    DummySystemEntityRecognizer.get_instance()
+                )
+        return SystemEntityRecognizer._instance
+
+    @abstractmethod
+    def parse(self, sentence, **kwargs):
+        """Calls System Entity Recognizer service API to extract numerical entities from a sentence.
+
+        Args:
+            sentence (str): A raw sentence.
+
+        Returns:
+            (tuple): A tuple containing:
+                - response (list, dict): Response from the System Entity Recognizer service that \
+                consists of a list of dicts, each corresponding to a single prediction or just a \
+                dict, corresponding to a single prediction.
+                - response_code (int): http status code.
+        """
+        pass
+
+    @abstractmethod
+    def resolve_system_entity(self, query, entity_type, span):
+        """Resolves a system entity in the provided query at the specified span.
+
+        Args:
+            query (Query): The query containing the entity
+            entity_type (str): The type of the entity
+            span (Span): The character span of the entity in the query
+
+        Returns:
+            Entity: The resolved entity
+
+        Raises:
+            SystemEntityResolutionError
+        """
+        pass
+
+    @abstractmethod
+    def get_candidates(self, query, entity_types=None, **kwargs):
+        """Identifies candidate system entities in the given query.
+
+        Args:
+            query (Query): The query to examine
+            entity_types (list of str): The entity types to consider
+
+        Returns:
+            list of QueryEntity: The system entities found in the query
+        """
+        pass
+
+    @abstractmethod
+    def get_candidates_for_text(self, text, entity_types=None, **kwargs):
+        """Identifies candidate system entities in the given text.
+
+        Args:
+            text (str): The text to examine
+            entity_types (list of str): The entity types to consider
+
+        Returns:
+            list of dict: The system entities found in the text
+        """
+        pass
+
+
+class DummySystemEntityRecognizer(SystemEntityRecognizer):
```

Comment:
Can we name this is `NoOpSystemEntityRecognizer`? It better reflects the function of this class.

Reply:
done